src/dotty/tools/dotc/core/OrderingConstraint.scala

```diff
@@ -0,0 +1,502 @@
+
+  /** The type variable corresponding to parameter numbered `n`, null if none was created */
+  private def typeVar(entries: Array[Type], n: Int): Type =
+    entries(paramCount(entries) + n)
+
+  /** The `boundsMap` entry corresponding to `param` */
+  def entry(param: PolyParam): Type = {
+    val entries = boundsMap(param.binder)
+    if (entries == null) NoType
+    else entries(param.paramNum)
+  }
+    
+// ----------- Contains tests --------------------------------------------------
+
+  def contains(pt: PolyType): Boolean = boundsMap(pt) != null
+
+  def contains(param: PolyParam): Boolean = {
+    val entries = boundsMap(param.binder)
+    entries != null && entries(param.paramNum).isInstanceOf[TypeBounds]
+  }
+
+  def contains(tvar: TypeVar): Boolean = {
+    val origin = tvar.origin
+    val entries = boundsMap(origin.binder)
+    val pnum = origin.paramNum
+    entries != null && isBounds(entries(pnum)) && (typeVar(entries, pnum) eq tvar)
+  }
+
+  private def isBounds(tp: Type) = tp.isInstanceOf[TypeBounds]
+  
+// ---------- Dependency handling ----------------------------------------------
+    
+  def lower(param: PolyParam): List[PolyParam] = lowerLens(this, param.binder, param.paramNum)
+  def upper(param: PolyParam): List[PolyParam] = upperLens(this, param.binder, param.paramNum)
+    
+  def minLower(param: PolyParam): List[PolyParam] = {
+    val all = lower(param)
+    all.filterNot(p => all.exists(isLess(p, _)))
+  }
+  
+  def minUpper(param: PolyParam): List[PolyParam] = {
+    val all = upper(param)
+    all.filterNot(p => all.exists(isLess(_, p)))
+  }
+  
+  def exclusiveLower(param: PolyParam, butNot: PolyParam): List[PolyParam] =
+    lower(param).filterNot(isLess(_, butNot))
+    
+  def exclusiveUpper(param: PolyParam, butNot: PolyParam): List[PolyParam] =
+    upper(param).filterNot(isLess(butNot, _))
+  
+// ---------- Info related to PolyParams -------------------------------------------
+
+  def isLess(param1: PolyParam, param2: PolyParam): Boolean =
+    upper(param1).contains(param2)
+
+  def nonParamBounds(param: PolyParam): TypeBounds = 
+    entry(param).asInstanceOf[TypeBounds]
+    
+  def fullLowerBound(param: PolyParam)(implicit ctx: Context): Type =
+    (nonParamBounds(param).lo /: minLower(param))(_ | _)
+
+  def fullUpperBound(param: PolyParam)(implicit ctx: Context): Type = 
+    (nonParamBounds(param).hi /: minUpper(param))(_ & _)
+    
+  def fullBounds(param: PolyParam)(implicit ctx: Context): TypeBounds =
+    nonParamBounds(param).derivedTypeBounds(fullLowerBound(param), fullUpperBound(param))
+
+  def typeVarOfParam(param: PolyParam): Type = {
+    val entries = boundsMap(param.binder)
+    if (entries == null) NoType
+    else {
+      val tvar = typeVar(entries, param.paramNum)
+      if (tvar != null) tvar else NoType
+    }
+  }  
+
+// ---------- Adding PolyTypes --------------------------------------------------
+   
+  /** The bound type `tp` without dependent parameters
```

Comment:
This comment should explain what a "dependent parameter" is, and what the argument `paramBuf` is for.